File: WebApi/godotrl/src/main/java/com/example/godotrl/cst/behavior/Domain.java
package com.example.godotrl.cst.behavior;

public final class Domain<T extends Number & Comparable<T>> implements Comparable<Domain<?>> {
    T val;

    public Domain(T val) {
        this.val = val;
    }

    @Override
    public int hashCode(){
        return this.val !=null ? this.val.hashCode() : 0;
    }

    @Override
    public int compareTo(Domain<?> state) {
        return doubleValue().compareTo(state.doubleValue());
        // return val.compareTo((T) state.val);

    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof Domain)) return false;
        return this.val != null && val.equals(((Domain<?>) obj).val);
    }

    public Integer intValue() {
        return (Integer) this.val;
    }

    public Long longValue() {
        return (Long) this.val;
    }

    public Float floatValue() {
        return (Float) this.val;
    }

    public Double doubleValue() {
        return (Double) this.val;
    }

    @Override
    public String toString() {
        return this.val.toString();
    }
}

File: GodotKotlin/MultiplatformFrogGame/src/main/kotlin/com/example/game/cst/behavior/Domain.kt
class Domain<T>(var `val`: T) : Comparable<Domain<*>?> where T : Number?, T : Comparable<T>? {
    @Override
    override fun hashCode(): Int {
        return if (`val` != null) `val`!!.hashCode() else 0
    }

    @Override
    operator fun compareTo(state: Domain<*>): Int {
        return `val`!!.compareTo(state.`val` as T)
    }

    @Override
    override fun equals(obj: Object): Boolean {
        if (this === obj) return true
        return if (obj !is Domain<*>) false else `val` != null && `val`!!.equals((obj as Domain<*>).`val`)
    }

    fun intValue(): Integer {
        return `val` as Integer
    }

    fun longValue(): Long {
        return `val` as Long
    }

    fun floatValue(): Float {
        return `val` as Float
    }

    fun doubleValue(): Double {
        return `val` as Double
    }

    @Override
    override fun toString(): String {
        return `val`.toString()
    }
}